### PR TITLE
Added plugins to run Checkstyle and generate reports

### LINKED ISSUE #3332: Add Java Checkstyle Plugin (1000USD Bounty)
---

**From the maintainer Li Haoyi**: I'm putting a 1000USD bounty on this issue, payable by bank transfer on a merged PR implementing this.

---

Checkstyle is a common Java linter, and we should support it as a `contrib/` module to better support Java projects. We should support all common configuration options, and provide it as an `CheckstyleModule` to allow usage on `JavaModule`s and `ScalaModule`s (which can contain Java sources). This needs to come with a reasonable set of unit tests, docs, and at least one example test that can be included in the documentation

Can possibly be adapted from https://github.com/etsy/sbt-checkstyle-plugin

### PR DESCRIPTION
**Pull request by ajaychandran**
Added plugins
- `CheckstyleModule`, fully compliant with [Checkstyle CLI](https://checkstyle.org/cmdline.html#Command_line_usage)
- `CheckstyleXsltModule`, generates reports by applying XSL Transformations on a Checkstyle output report

Resolves #3332.

### CODE REVIEW

**Comment by ajaychandran** [comment]
**Pending**
On running `mill.contrib.checkstyle.CheckstyleModuleTest.arguments.stdout`, the Checkout output report is not printed to the console.
```
-------------------------------- Running Tests --------------------------------
mill.testkit.UnitTester: [8/8] checkstyle 
mill.testkit.UnitTester: running checkstyle ...
Checkstyle ends with 4 errors.
mill.testkit.UnitTester: checkstyle found 4 violation(s)
+ mill.contrib.checkstyle.CheckstyleModuleTest.arguments.stdout  
```
I suspect this is a `classpath` issue. Will investigate further.

**Comment by ajaychandran** [comment]
@lihaoyi Please review.

File: contrib/checkstyle/src/mill/contrib/checkstyle/CheckstyleXsltModule.scala [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+   *
+   * html/xslt0.xml -> xslt0.html
+   * html/xslt1.xml -> xslt1.html
+   * pdf/xslt1.xml  -> xslt1.pdf
+   * pdf/xslt2.xml  -> xslt2.pdf
+   *
+   * }}}
+   */
+  def checkstyleXsltReports: T[Set[CheckstyleXsltReport]] = T {
+    val dir = millSourcePath / "checkstyle-xslt"
```
Comment (lihaoyi) [inline contrib/checkstyle/src/mill/contrib/checkstyle/CheckstyleXsltModule.scala:73]: This should be moved into a `T.source` task to ensure changes to that upstream folder properly invalidate this downstream target

**Comment by lihaoyi** [comment]
I think this looks good, if you're happy with it I'll merge it. Another CheckStyle PR has been submitted that has also been of great help understanding the solution space (https://github.com/com-lihaoyi/mill/pull/3472), so I'll pay out 70% of the bounty for this one and 30% of the bounty for the other. @ajaychandran email me your bank transfer details and I'll wire it over

**Comment by ajaychandran** [comment]
@lihaoyi Thank you for the guidance.

File: contrib/checkstyle/src/mill/contrib/checkstyle/CheckstyleXsltModule.scala (hunk 1 above)
Comment (ajaychandran) [reply contrib/checkstyle/src/mill/contrib/checkstyle/CheckstyleXsltModule.scala:73]: Remove `CheckstyleXsltReport` and replace `checkstyleXsltReports` with
```scala

def checkstyleXslt: T[Seq[PathRef]] = T.sources { ??? }

def checkstyleXsltOutput: T[Seq[PathRef]] = T { ??? }
```
I can submit a PR. Let me know.

Comment (lihaoyi) [reply contrib/checkstyle/src/mill/contrib/checkstyle/CheckstyleXsltModule.scala:73]: no worries, I already opened on https://github.com/com-lihaoyi/mill/pull/3522

**Pull request merged**